### DISCUSSION TITLE
EmailProvider + Adapter

### DISCUSSION THREAD

**Discussion by robahtou** [nextauthjs/next-auth#8668] [Help]
### Summary

I'm working on creating my own Adapter using the EmailProvider. 
I have a very narrow task I'd like to accomplish with the EmailProvider. 

Instead of having a login form with `username` & `password`, I'd like to have a form that only has one input: `email`.
When a user inputs an email, I'd like to send them a magiclink to confirm their email.  Here is what my auth file looks like with the EmailProvider (I am using Resend as my email provider)

`/api/auth/[...nextauth]/route.ts`
```
providers: [
 EmailProvider({
   server: {
     host: process.env.EMAIL_SERVER_HOST,
     port: process.env.EMAIL_SERVER_PORT,
     auth: {
       user: process.env.EMAIL_SERVER_USER,
       pass: process.env.EMAIL_SERVER_PASSWORD
     }
   },
   from: 'no-reply@example.com',
   maxAge: 15 * 60,
   async generateVerificationToken() {
     return 'ABC123';
   },
   async sendVerificationRequest(params) {
     console.log({ params });
     const { identifier, url, provider, theme } = params;

     const result = await resend.emails.send({
       from: provider.from,
       to: [identifier],
       subject: 'verify email',
       react: EmailTemplate({ firstName: 'John', url }),
     });

     console.log('result', result);
   }
 })
 ],
 adapter: MyAdapter({
   url: process.env.MY_ADAPTOR_URL
   secret: process.env.MY_ADAPTOR_SECRET
 }),
```

For the adapter, I copied the example code from here: https://next-auth.js.org/tutorials/creating-a-database-adapter

The above works to the extent that I can receive an email with a link in it. But when I click on that link I always get this error message: 
```
[next-auth][error][CALLBACK_EMAIL_ERROR]
https://next-auth.js.org/errors#callback_email_error Cannot read properties of undefined (reading 'valueOf') {
  message: "Cannot read properties of undefined (reading 'valueOf')",
  stack: "TypeError: Cannot read properties of undefined (reading 'valueOf')\n" +
    '    at Object.callback (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/routes/callback.js:179:61)\n' +
    '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
    '    at async AuthHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/index.js:202:38)\n' +
    '    at async NextAuthRouteHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/next/index.js:50:30)\n' +
    '    at async authOptions (webpack-internal:///(rsc)/./src/app/api/auth/[...nextauth]/route.ts:22:12)\n' +
    '    at async eval (webpack-internal:///(rsc)/./node_modules/.pnpm/next@13.4.19_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/future/route-modules/app-route/module.js:254:37)',
  name: 'TypeError'
}
```
I check the error codes for `CALLBACK_EMAIL_ERROR` and find this https://next-auth.js.org/errors#callback_email_error:
```
This can occur during the email callback process. Specifically, if there was an error signing the user in via email, encoding the jwt, etc.

Please double check your Email settings.
```

What exactly should I be looking for in the Email settings? Is there something wrong with the above route configuration?

### Additional information

```js
Error message: `CALLBACK_EMAIL_ERROR`

[next-auth][error][CALLBACK_EMAIL_ERROR]
https://next-auth.js.org/errors#callback_email_error Cannot read properties of undefined (reading 'valueOf') {
  message: "Cannot read properties of undefined (reading 'valueOf')",
  stack: "TypeError: Cannot read properties of undefined (reading 'valueOf')\n" +
    '    at Object.callback (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/routes/callback.js:179:61)\n' +
    '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
    '    at async AuthHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/index.js:202:38)\n' +
    '    at async NextAuthRouteHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/next/index.js:50:30)\n' +
    '    at async authOptions (webpack-internal:///(rsc)/./src/app/api/auth/[...nextauth]/route.ts:22:12)\n' +
    '    at async eval (webpack-internal:///(rsc)/./node_modules/.pnpm/next@13.4.19_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/future/route-modules/app-route/module.js:254:37)',
  name: 'TypeError'
}
```
```


### Example

_No response_

**Comment by AnozieChibuike** [comment]
I am having the same issues when I defined my custom adapter

**Comment by AnozieChibuike** [comment]
> ### Summary
> I'm working on creating my own Adapter using the EmailProvider. I have a very narrow task I'd like to accomplish with the EmailProvider.
> 
> Instead of having a login form with `username` & `password`, I'd like to have a form that only has one input: `email`. When a user inputs an email, I'd like to send them a magiclink to confirm their email. Here is what my auth file looks like with the EmailProvider (I am using Resend as my email provider)
> 
> `/api/auth/[...nextauth]/route.ts`
> 
> ```
> providers: [
>  EmailProvider({
>    server: {
>      host: process.env.EMAIL_SERVER_HOST,
>      port: process.env.EMAIL_SERVER_PORT,
>      auth: {
>        user: process.env.EMAIL_SERVER_USER,
>        pass: process.env.EMAIL_SERVER_PASSWORD
>      }
>    },
>    from: 'no-reply@example.com',
>    maxAge: 15 * 60,
>    async generateVerificationToken() {
>      return 'ABC123';
>    },
>    async sendVerificationRequest(params) {
>      console.log({ params });
>      const { identifier, url, provider, theme } = params;
> 
>      const result = await resend.emails.send({
>        from: provider.from,
>        to: [identifier],
>        subject: 'verify email',
>        react: EmailTemplate({ firstName: 'John', url }),
>      });
> 
>      console.log('result', result);
>    }
>  })
>  ],
>  adapter: MyAdapter({
>    url: process.env.MY_ADAPTOR_URL
>    secret: process.env.MY_ADAPTOR_SECRET
>  }),
> ```
> 
> For the adapter, I copied the example code from here: https://next-auth.js.org/tutorials/creating-a-database-adapter
> 
> The above works to the extent that I can receive an email with a link in it. But when I click on that link I always get this error message:
> 
> ```
> [next-auth][error][CALLBACK_EMAIL_ERROR]
> https://next-auth.js.org/errors#callback_email_error Cannot read properties of undefined (reading 'valueOf') {
>   message: "Cannot read properties of undefined (reading 'valueOf')",
>   stack: "TypeError: Cannot read properties of undefined (reading 'valueOf')\n" +
>     '    at Object.callback (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/routes/callback.js:179:61)\n' +
>     '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
>     '    at async AuthHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/index.js:202:38)\n' +
>     '    at async NextAuthRouteHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/next/index.js:50:30)\n' +
>     '    at async authOptions (webpack-internal:///(rsc)/./src/app/api/auth/[...nextauth]/route.ts:22:12)\n' +
>     '    at async eval (webpack-internal:///(rsc)/./node_modules/.pnpm/next@13.4.19_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/future/route-modules/app-route/module.js:254:37)',
>   name: 'TypeError'
> }
> ```
> 
> I check the error codes for `CALLBACK_EMAIL_ERROR` and find this https://next-auth.js.org/errors#callback_email_error:
> 
> ```
> This can occur during the email callback process. Specifically, if there was an error signing the user in via email, encoding the jwt, etc.
> 
> Please double check your Email settings.
> ```
> 
> What exactly should I be looking for in the Email settings? Is there something wrong with the above route configuration?
> 
> ### Additional information
> ```js
> Error message: `CALLBACK_EMAIL_ERROR`
> 
> [next-auth][error][CALLBACK_EMAIL_ERROR]
> https://next-auth.js.org/errors#callback_email_error Cannot read properties of undefined (reading 'valueOf') {
>   message: "Cannot read properties of undefined (reading 'valueOf')",
>   stack: "TypeError: Cannot read properties of undefined (reading 'valueOf')\n" +
>     '    at Object.callback (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/routes/callback.js:179:61)\n' +
>     '    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)\n' +
>     '    at async AuthHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/core/index.js:202:38)\n' +
>     '    at async NextAuthRouteHandler (webpack-internal:///(rsc)/./node_modules/.pnpm/next-auth@4.23.1_next@13.4.19_nodemailer@6.9.5_react-dom@18.2.0_react@18.2.0/node_modules/next-auth/next/index.js:50:30)\n' +
>     '    at async authOptions (webpack-internal:///(rsc)/./src/app/api/auth/[...nextauth]/route.ts:22:12)\n' +
>     '    at async eval (webpack-internal:///(rsc)/./node_modules/.pnpm/next@13.4.19_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/future/route-modules/app-route/module.js:254:37)',
>   name: 'TypeError'
> }
> ```
> 
> ```
> 
> 
> ### Example
> 
> _No response_
> ```

Could you fix it?

**Comment by AnozieChibuike** [comment]
@robahtou  So I spent today looking to fix the error, I had to look into the next-auth/core source code, So the error comes from `useVerificationToken` function of the customAdapter, you need to define a new key called `expires`  (Which obviously if you were retrieving from a db would exist), setting it to a date object... the valueOF error you saw was nextauth core comparing an undefined value(the missing `expires` key) to a date 
![image](https://github.com/nextauthjs/next-auth/assets/100452992/fbfb3708-ef07-4528-9c60-660efeede066)
